numpy/lib/arraypad.py

```diff
@@ -1193,192 +786,128 @@ def pad(array, pad_width, mode, **kwargs):
-
-    elif mode == 'median':
-        for axis, ((pad_before, pad_after), (chunk_before, chunk_after)) \
-                in enumerate(zip(pad_width, kwargs['stat_length'])):
-            newmat = _prepend_med(newmat, pad_before, chunk_before, axis)
-            newmat = _append_med(newmat, pad_after, chunk_after, axis)
-
-    elif mode == 'minimum':
-        for axis, ((pad_before, pad_after), (chunk_before, chunk_after)) \
-                in enumerate(zip(pad_width, kwargs['stat_length'])):
-            newmat = _prepend_min(newmat, pad_before, chunk_before, axis)
-            newmat = _append_min(newmat, pad_after, chunk_after, axis)
-
-    elif mode == 'reflect':
-        for axis, (pad_before, pad_after) in enumerate(pad_width):
-            if narray.shape[axis] == 0:
-                # Axes with non-zero padding cannot be empty.
-                if pad_before > 0 or pad_after > 0:
-                    raise ValueError("There aren't any elements to reflect"
-                                     " in axis {} of `array`".format(axis))
-                # Skip zero padding on empty axes.
-                continue
-
-            # Recursive padding along any axis where `pad_amt` is too large
-            # for indexing tricks. We can only safely pad the original axis
-            # length, to keep the period of the reflections consistent.
-            if ((pad_before > 0) or
-                    (pad_after > 0)) and newmat.shape[axis] == 1:
+    }
+    unsupported_kwargs = set(kwargs) - set(allowed_kwargs[mode])
+    if unsupported_kwargs:
+        raise ValueError("unsupported keyword arguments for mode '{}': {}"
+                         .format(mode, unsupported_kwargs))
+
+    stat_functions = {"maximum": np.max, "minimum": np.min,
+                      "mean": np.mean, "median": np.median}
+
+    # Create array with final shape and original values
+    # (padded area is undefined)
+    padded, _ = _pad_simple(array, pad_width)
+    # And prepare iteration over all dimensions
+    # (zipping may be more readable than using enumerate)
+    axes = range(padded.ndim)
+
+    if array.size == 0 and mode not in {"constant", "empty"}:
+        # Deal with special case: only modes "constant" and "empty" can extend
+        # empty axes, all other modes depend on `array` not being empty.
+        for axis, index_pair in zip(axes, pad_width):
+            if (
+                array.shape[axis] == 0
+                and (index_pair[0] > 0 or index_pair[1] > 0)
+            ):
+                raise ValueError(
+                    "can't extend empty axis {} using modes other than "
+                    "'constant' or 'empty'".format(axis)
+                )
+        # If pad values in empty axis were zero as well, then _pad_simple
+        # already returned the correct result
+        return padded
+
+    if mode == "constant":
+        values = kwargs.get("constant_values", 0)
+        values = _as_pairs(values, padded.ndim, assert_number=True)
```

Comment:
Does this prohibit padding non-numeric arrays?

Reply:
Yes, but now I see that the old implementation can deal with non-numeric arrays as well. I'll fix that.

Reply:
I guess a test needs to be added to preserve the behaviour in the future.

Reply:
@hmaarrfk: A great way for you to contribute without stepping on each others toes here would be to write some of those tests in a separate PR on master

Reply:
:sweat_smile: of course! sorry about that. https://github.com/numpy/numpy/pull/11958